### ISSUE TITLE
MMvec refactor

### ISSUE THREAD

**Issue by mortonjt** [biocore/mmvec#166]
We're going to go [pytorch](https://www.pytorchlightning.ai/) OR [numpyro](https://github.com/pyro-ppl/numpyro).  The framework will have the following skeleton

`model.py (mmvec.py)`

```python
import torch
import torch.nn
from torch.distributions import Multinomial

class MMvec(nn.Module):
    def __init__(self, num_microbes, num_metabolites, latent_dim):
        self.encoder = nn.Embedding(num_microbes, latent_dim)
        self.decoder = nn.Sequential([nn.Linear(latent_dim, num_metabolite), nn.Softmax()])
        # TODO : may want to have a better softmax

    def forward(X, Y):
        """ X is one-hot encodings (B x num_microbes).  Y is metabolite abundances (B x num_metabolites).  B is the batch size""" 
        z = self.encoder(X)
        pred_y = self.decoder(z)
        lp = Multinomial(pred_y).log_prob(Y).sum()
        return lp
```

`train.py` (could use Pytorch lightning)

The wishlist
- Early stopping (see [video](https://pytorch-lightning.readthedocs.io/en/stable/common/early_stopping.html) for example)
- [Arviz](https://github.com/arviz-devs/arviz) for diagnostics diagnostics
- Typing would be great.  See [torchtyping](https://github.com/patrick-kidger/torchtyping)
- Torchtests could be cool also.  See [torchtest](https://github.com/suriyadeepan/torchtest)
- Being Bayesian would be nice.  [SWAG](https://pytorch.org/blog/pytorch-1.6-now-includes-stochastic-weight-averaging/) is the laziest approach

**Comment by Keegan-Evans** [comment]
First pass:

https://github.com/biocore/mmvec/pull/167

**Comment by Keegan-Evans** [comment]
First pass:

https://github.com/Keegan-Evans/mmvec/tree/pytorch-refactor

**Comment by mortonjt** [comment]
Hi @Keegan-Evans this is a great first pass.  The basic architecture is there, and it looks like the gradient descent is working.

There are a couple of things that we'll want to try out
1. Getting the unittests to pass at https://github.com/biocore/mmvec/blob/master/mmvec/tests/test_multimodal.py#L18
2. Doubling checking the soils experiment at https://github.com/biocore/mmvec/blob/master/mmvec/tests/test_multimodal.py#L76

We may want to revisit the decoder architecture -- the softmax identifability may end up biting us when interpreting the decoder parameters (probably also why you see word2vec applications interpreting only the encoder parameters, rather than decoder parameters).   In MMvec, we used the inverse ALR transform, which would look something like

```python
class LinearALR(nn.Module):
    def __init__(self, input_dim, output_dim):
        W = nn.Parameter(torch.randn(output_dim - 1, input_dim))
        b = nn.Parameter(torch.randn(output_dim - 1))

    def forward(self, x):
        b = x.shape[0]
        z = torch.zeros((b, 1))
        x = torch.stack((z, x), axis=1)
        y = W @ x + b
        return F.softmax(y)
```

ALR also has issues, we may want to use ILR -- but I'll need some time to write that out.

**Comment by mortonjt** [comment]
Here is a statistical significance test for differential abundance

https://github.com/flatironinstitute/q2-matchmaker/blob/main/q2_matchmaker/_stats.py#L46

**Comment by Keegan-Evans** [comment]
@mortonjt,

In your original model, when constructing the ranks, you stack zeros onto the product of U and V [here](https://github.com/biocore/mmvec/blob/master/mmvec/multimodal.py#L214)? Is it necessary or expedient for how you performed the ALR?

**Comment by mortonjt** [comment]
Hi @Keegan-Evans this is necessary if the V matrix is in ALR coordinates.  If we're using ILR, then it would be `U @ V @ Psi.T` (if `Psi` is D x D - 1)

**Comment by Keegan-Evans** [comment]
@mortonjt, I was working on the ALR this morning and realized that might be the case, thanks for the reply!
